docs/specs/short-read-alignment-vector.md

```diff
@@ -0,0 +1,215 @@
+
+* Any demultiplexing of multiple samples within a single sequencing
+  lane will have been resolved upstream of this pipeline. I.e., the
+  pipeline will run from demultiplexed sequence reads. The term
+  "lanelet" is used here to mean data from a single lane of sequencing
+  from a single sample.
+
+* Each sample may have data from multiple sequencing runs (lanes), and
+  the number of lanelets may vary between samples. I.e., the pipeline
+  cannot assume a fixed number of lanelets per sample.
+
+* The complete set of input data will be known ahead of time, before
+  the pipeline is launched for a given batch of samples. I.e., there
+  is no need to consider any case where additional lanelets may arrive
+  at a later time.
+
+* Sequencing will be from Illumina technology, using paired-end
+  sequencing. Mosquitoes will be sequenced individually, i.e., each
+  mosquito sample will be used to create a separate DNA library, and
+  libraries will not be pooled. Each individual mosquito will be
+  sequenced to high coverage (target ~30X). Read length may vary
+  between sequencing runs, i.e., the pipeline cannot assume the same
+  read length across all lanelets.
+
+* Trimming of adapter sequences, if necessary, will have been
+  performed prior to running this pipeline.
+
+
+## Inputs
+
+* Reference sequence - The reference genome sequence to align to, in
+  FASTA format.
+
+* Sequence reads - For each lanelet, a pair of files containing
+  sequence reads, with one file containing forward reads and one file
+  containing reverse reads, in FASTQ format.
+
+* Lanelet manifest - A tab delimited file with one row per
+  lanelet. Mandatory columns are as follows, may contain any other
+  columns:
+
+    * `read1_path` - URL or local filesystem path to FASTQ file
+      containing forward reads.
+
+    * `read2_path` - URL or local filesystem path to FASTQ file
+      containing reverse reads.
+
+    * `sample_id` - Identifier for the sample that was sequenced.
+
+
+## Outputs
+
+* One analysis-ready BAM file for each sample.
+
+* samtools stats output for each sample.
+
+* GATK CallableLoci output for each sample.
+
+
+## Software
+
+* bwa 0.7.15
+* samtools 1.4.1 (htslib 1.4.1)
+* picard 2.9.2
+* biobambam 2.0.73
+* GATK 3.7-0
+
+
+## Pipeline
+
+
+### Sub-pipeline: Lanelet alignment
+
+For each lanelet, the following steps are run.
+
+
+#### Step: Read alignment
+
+Align sequence reads for a single lanelet to the reference sequence with command:
+
+```
+bwa mem -M -T 0 -R readgroup_string ref.fa read1.fastq read2.fastq > output.sam
+```
+
+Notes:
+
+* -M marks split alignments as secondary (for Picard compatibility)
+
+* -T 0 sets the threshold for alignment score output to zero
+
+
+#### Step: Read alignment post-processing
+
+Conversion of sam to bam via namesort, fixmates, coordinate sort and
+calmd:
+
+```
+samtools view -bu output.sam 
+| samtools sort -n 
+| samtools fixmate 
+| samtools sort  
+| samtools calmd -b ref.fa > output.bam
```

Comment:
I was googling calmd to remind em what it is, and noticed this line in the [docs.](http://www.htslib.org/doc/samtools-calmd.html)

`Calmd can also read and write CRAM files although in most cases it is pointless as CRAM recalculates MD and NM tags on the fly. The one exception to this case is where both input and output CRAM files have been / are being created with the no_ref option.`

If I'm reading this correctly, it made me wonder if this calmd step could be skipped if we decided to move to crams instead?

Reply:
From personal correspondance with Laura from GATK, there was concern about java libraries for reading/converting crams, and that's why the GATK 4 best practices always use bams.  However, I don't know which versions and which libraries (HTSDK?).  We can follow up to determine whether it's an issue in GATK 3 or GATK 4.

Reply:
These specs were first designed with reuse of legacy tools in mind.  Given that we don't have these legacy constraints, perhaps we can be flexible with the steps. But we may end up with differing output from previous mosquito sequencing data releases.

Reply:
There are probably two questions here. (1) Switch to use of CRAM within the pipeline for data exchange between steps (maybe saves disk use for transient files during pipeline execution)? (2) Switch to CRAM as final output of the pipeline (maybe saves disk use for persistent outputs)?

Re (1), if at all possible I'd like to go for an implementation that sticks as closely as possible to this spec, so we get data that is comparable to what we have built previously. So I'd say let's avoid making any changes to internal steps within the pipeline, given that we know this spec works.

Re (2) I'd be happy for an implementation to add a BAM -> CRAM conversion step at the end, if it becomes clear that we need to save disk space. But that can probably be an implementation detail, not necessary to specify here.